examples/tcp-wait/client.rs

```diff
@@ -0,0 +1,265 @@
+
+        // Ensure that scatter-gather array has the requested size.
+        assert!(sga.sga_segs[0].sgaseg_len as usize == size);
+
+        // Fill in scatter-gather array.
+        let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
+        let len: usize = sga.sga_segs[0].sgaseg_len as usize;
+        let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+        slice.fill(value);
+
+        Ok(sga)
+    }
+
+    fn issue_push(&mut self, i: usize, sockqd: QDesc) -> Result<QToken, anyhow::Error> {
+        let fill_char: u8 = (i % (u8::MAX as usize - 1) + 1) as u8;
+        const BUFSIZE: usize = 64;
+        let sga: demi_sgarray_t = self.make_sgarray(BUFSIZE, fill_char)?;
+        let qt: QToken = self.libos.push(sockqd, &sga)?;
+        Ok(qt)
+    }
+
+    fn connect_to_server(&mut self, num_clients: usize) -> Result<QDesc, anyhow::Error> {
+        let sockqd: QDesc = self.libos.socket(AF_INET, SOCK_STREAM, 0)?;
+        let qt: QToken = self.libos.connect(sockqd, self.remote)?;
+        let qr: demi_qresult_t = self.libos.wait(qt, None)?;
+        match qr.qr_opcode {
+            demi_opcode_t::DEMI_OPC_CONNECT => {
+                println!("{} clients connected", num_clients + 1);
+            },
+            demi_opcode_t::DEMI_OPC_FAILED => {
+                // Close socket on error.
+                // FIXME: https://github.com/demikernel/demikernel/issues/640
+                anyhow::bail!("operation failed (qr_ret={:?})", qr.qr_ret)
+            },
+            qr_opcode => {
+                // Close socket on error.
+                // FIXME: https://github.com/demikernel/demikernel/issues/640
+                anyhow::bail!("unexpected result (qr_opcode={:?})", qr_opcode)
+            },
+        }
+        Ok(sockqd)
+    }
+
+    // Attempt to wait for a push() operation to complete after asynchronous closing a socket.
+    pub fn push_async_close_wait(&mut self) -> Result<()> {
+        for i in 0..self.nclients {
+            let sockqd: QDesc = self.connect_to_server(i)?;
+
+            // Close socket on error.
+            // FIXME: https://github.com/demikernel/demikernel/issues/640
+            let push_qt: QToken = self.issue_push(i, sockqd)?;
+            let async_close_qt: QToken = self.libos.async_close(sockqd)?;
+
+            // Wait for async_close().
+            match self.libos.wait(async_close_qt, None) {
+                Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_CLOSE && qr.qr_ret == 0 => {},
+                Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_FAILED && qr.qr_ret == libc::ECONNRESET => {},
+                Ok(_) => anyhow::bail!("wait() should succeed with async_close()"),
+                Err(_) => anyhow::bail!("wait() should succeed with async_close()"),
+            }
+
+            // Wait for push().
+            match self.libos.wait(push_qt, None) {
+                Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_PUSH && qr.qr_ret == 0 => {},
```

Comment:
You should assert for `ECANCELED` as well.

Reply:
Done.